app/forms/idv/api_image_upload_form.rb

```diff
@@ -0,0 +1,78 @@
+module Idv
+  class ApiImageUploadForm
+    include ActiveModel::Model
+    include ActionView::Helpers::TranslationHelper
+
+    IMAGE_KEYS = %i[
+      front
+      back
+      selfie
+    ].freeze
+
+    validates_presence_of :front
+    validates_presence_of :back
+    validates_presence_of :selfie, if: :liveness_checking_enabled?
+
+    validate :validate_images
+
+    def initialize(params, liveness_checking_enabled:)
+      @params = params
+      @liveness_checking_enabled = liveness_checking_enabled
+    end
+
+    # Normally we'd return FormResponse, but that has errors as a hash,
+    # where the proofer reponses have errors as an array. This is easier to compare
+    # with the proofer responses
+    # @return [DocAuthClient::Response]
+    def submit
+      DocAuthClient::Response.new(
+        success: valid?,
+        errors: errors.full_messages,
+      )
+    end
+
+    def liveness_checking_enabled?
+      @liveness_checking_enabled
+    end
+
+    IMAGE_KEYS.each do |image_key|
+      define_method(image_key) do
+        params[image_key]
+      end
+    end
+
+    def self.human_attribute_name(attr, options = {})
+      # i18n-tasks-use t('doc_auth.headings.front')
+      # i18n-tasks-use t('doc_auth.headings.back')
+      # i18n-tasks-use t('doc_auth.headings.selfie')
+      I18n.t(attr, options.merge(scope: 'doc_auth.headings'))
+    end
```

Comment:
So for the sake of simplicity, I used the translation strings for the frontend's headers for the fields. But when used in actual error strings it makes for some kind of weird sentences that might not be that clear:

> Front image data URL not formatted correctly

Reply:
Maybe we should add keys for the descriptive field names? Such as `Picture of front of ID`?

On the other hand, an error like this indicates programmer error, so it's not something a user would see often in the wild, may not be worth the effort

Reply:
I think it might depend on how we want the design of the errors being presented, but my expectation was that the errors returned wouldn't be full text strings like "Front image is too blurry", but rather a generic error message "Image is too blurry" associated with a particular key of the endpoint `front`. That way, as long as the front-end can map that key back to the field where it's shown, it can present the error in a way that's contextually relevant to the user, even if the text of the error itself is generic.

I could see the full error messages being useful if we wanted to show a combined list of all of the problems with the form submission.

![image](https://user-images.githubusercontent.com/1779930/90258757-fcca8900-de16-11ea-8952-5781d96396d4.png)

As for how this would be structured, maybe something like:

```json
{
  "success": false,
  "errors": [
    { "field": "front", "error": "Image is too blurry" }
  ]
}
```

Reply:
So we're in a bit of a pickle. I agree that being able to label fields and errors like that would be great, but it turns out that our code that parses responses from Acuant and gives us errors, it returns a flat array of strings, with no detail about which field they come from.

Acuant gives us English strings. We match those to keys we have, then we re-translate those into our supported locales:

https://github.com/18F/identity-idp/blob/master/config/friendly_error/config.yml

So maybe we could add some metadata and select a field for each of these messages we want to highlight? But also this is just one vendor's errors and I'm not sure what our future vendor's error formats would look like

Reply:
> So we're in a bit of a pickle. I agree that being able to label fields and errors like that would be great, but it turns out that our code that parses responses from Acuant and gives us errors, it returns a flat array of strings, with no detail about which field they come from.

Trying to follow this. I see that our `AcuantClient` will return errors that aren't associated with a particular field, but it seems like this would be within our control to change? In other words, rather than returning just the response data for the failed submission, return both the response data merged with the field which had been submitted at the time the error occurs.

https://github.com/18F/identity-idp/blob/3c645fa7ef8f31632710840d711045b58b5b34fc/app/services/acuant/acuant_client.rb#L40-L60

Something like...

```rb
      front_image_response = post_front_image(image: front_image, instance_id: instance_id)
      return [ :front_image, front_image_response ] unless front_image_response.success?
```

Would that help? Or am I on the wrong track here?

Reply:
Yeah we can do something along those lines... unfortunate the `post_images` method gets called in our existing `DocumentCaptureStep` and `MobileCaptureStep` as part of our flow so I'd need to make sure that those steps would be able to handle a type change like that.

Can we follow up about error formats in another PR?

Reply:
> Can we follow up about error formats in another PR?

Sure 👍